p2p/src/authenticated/actors/dialer.rs

```diff
@@ -134,17 +144,47 @@ impl<E: Spawner + Clock + GClock + Network + Rng + CryptoRng + Metrics, C: Schem
     async fn run(
         mut self,
         mut tracker: tracker::Mailbox<E, C>,
-        mut supervisor: spawner::Mailbox<E, SinkOf<E>, StreamOf<E>, C>,
+        mut supervisor: spawner::Mailbox<E, SinkOf<E>, StreamOf<E>, C::PublicKey>,
     ) {
+        let mut dial_deadline = self.context.current();
+        let mut query_deadline = self.context.current();
         loop {
-            // Attempt to dial peers we know about
-            self.dial_peers(&mut tracker, &mut supervisor).await;
-
-            // Sleep for a random amount of time up to the dial frequency
-            let wait = self
-                .context
-                .gen_range(Duration::default()..self.dial_frequency);
-            self.context.sleep(wait).await;
+            select! {
+                _ = self.context.sleep_until(dial_deadline) => {
+                    // Update the deadline.
+                    dial_deadline = self.random_wait(self.dial_frequency);
+
+                    // Pop the queue until we can reserve a peer.
+                    // If a peer is reserved, attempt to dial it.
+                    while let Some(peer) = self.queue.pop() {
+                        // Attempt to reserve peer.
+                        let Some(reservation) = tracker.dial(peer).await else {
+                            continue;
+                        };
+                        self.dial_peer(reservation, &mut supervisor).await;
+                    }
+                },
+                _ = self.context.sleep_until(query_deadline) => {
+                    // Update the deadline.
+                    query_deadline = self.random_wait(self.query_frequency);
```

Comment:
May be worth noting that this approach "should" make it more likely that our connection rate limiter isn't maxxed out for a given peer from dial exhaustion (making it more likely they can connect to us when their dialer fires).